src/mbgl/map/vector_tile.hpp

```diff
@@ -1,117 +1,55 @@
+class VectorTileFeature : public GeometryTileFeature {
 public:
-    VectorTileFeature(pbf feature, const VectorTileLayer& layer);
+    VectorTileFeature(pbf, const VectorTileLayer&);
 
+    virtual FeatureType getType() const { return type; }
+    virtual mapbox::util::optional<Value> getValue(const std::string&) const;
+    virtual GeometryCollection getGeometries() const;
+
+private:
     uint64_t id = 0;
     FeatureType type = FeatureType::Unknown;
     std::map<std::string, Value> properties;
-    pbf geometry;
+    pbf geometry_pbf;
 };
 
-std::ostream& operator<<(std::ostream&, const VectorTileFeature& feature);
-
-
-class VectorTileTagExtractor {
+class VectorTileLayer : public GeometryTileLayer {
 public:
-    VectorTileTagExtractor(const VectorTileLayer &layer);
+    VectorTileLayer(pbf);
 
-    void setTags(const pbf &pbf);
-    mapbox::util::optional<Value> getValue(const std::string &key) const;
-    void setType(FeatureType type);
-    FeatureType getType() const;
+    virtual std::size_t featureCount() const { return features.size(); }
+    virtual const util::ptr<const GeometryTileFeature> feature(std::size_t) const;
 
 private:
-    const VectorTileLayer &layer_;
-    pbf tags_;
-    FeatureType type_ = FeatureType::Unknown;
-};
-
-/*
- * Allows iterating over the features of a VectorTileLayer using a
- * BucketDescription as filter. Only features matching the descriptions will
- * be returned (as pbf).
- */
-class FilteredVectorTileLayer {
-public:
-    class iterator {
-    public:
-        iterator(const FilteredVectorTileLayer& filter, const pbf& data);
-        void operator++();
-        bool operator!=(const iterator& other) const;
-        const pbf& operator*() const;
-
-    private:
-        const FilteredVectorTileLayer& parent;
-        bool valid = false;
-        pbf feature;
-        pbf data;
-    };
+    friend class VectorTile;
+    friend class VectorTileFeature;
 
-public:
-    FilteredVectorTileLayer(const VectorTileLayer& layer, const FilterExpression &filterExpression);
-
-    iterator begin() const;
-    iterator end() const;
-
-private:
-    const VectorTileLayer& layer;
-    const FilterExpression& filterExpression;
-};
-
-std::ostream& operator<<(std::ostream&, const PositionedGlyph& placement);
-
-class VectorTileLayer {
-public:
-    VectorTileLayer(pbf data);
-
-    const pbf data;
     std::string name;
     uint32_t extent = 4096;
     std::vector<std::string> keys;
-    std::unordered_map<std::string, uint32_t> key_index;
     std::vector<Value> values;
-    std::map<std::string, std::map<Value, Shaping>> shaping;
+    std::vector<pbf> features;
 };
 
-class VectorTile {
+class VectorTile : public GeometryTile {
 public:
-    VectorTile();
-    VectorTile(pbf data);
-    VectorTile& operator=(VectorTile&& other);
-
-    std::map<std::string, const VectorTileLayer> layers;
-};
+    VectorTile(pbf);
 
+    virtual const util::ptr<const GeometryTileLayer> getLayer(const std::string&) const;
```

Comment:
This is not a virtual function, but it overrides a virtual function. Instead of specifying it as `virtual`, please use [`override`](http://en.cppreference.com/w/cpp/language/override).